idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ReplaceCallFixFactories.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.fixes
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.frontend.api.symbols.markers.KtTypedSymbol
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeNullability
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeWithNullability
+import org.jetbrains.kotlin.idea.quickfix.ReplaceImplicitReceiverCallFix
+import org.jetbrains.kotlin.idea.quickfix.ReplaceWithSafeCallFix
+import org.jetbrains.kotlin.idea.quickfix.quickFixesPsiBasedFactoryWithAnalysis
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.psi.*
+
+object ReplaceCallFixFactories {
+    val UnsafeCallFactory = quickFixesPsiBasedFactoryWithAnalysis { psiElement: PsiElement ->
+        fun KtExpression.shouldHaveNotNullType(): Boolean {
+            // This function is used to determine if we may need to add an elvis operator after the safe call. For example, to replace
+            // `s.length` in `1 + s.length` with a safe call, it should be replaced with `(s.length ?: <caret>)`.
+            // NOTE: Even though we could, we don't check to see if the call is an argument and if the parameter type is nullable.
+            // FE1.0 doesn't either (see ReplaceWithSafeCallFixFactory).
+            val type = when (val parent = parent) {
+                is KtBinaryExpression -> parent.left?.getKtType()
+                is KtProperty -> {
+                    // Case: `val i = s.length`. `parent.getReturnKtType()` would return Int and therefore an elvis would be added,
+                    // which does not match FE1.0 behavior. We want to see if there is an explicit type, e.g., `val i: Int = s.length`
+                    // and default to the replacement being nullable if there is no explicit type.
+                    if (parent.typeReference != null) {
+                        parent.getReturnKtType()
+                    } else null
+                }
+                else -> null
+            }
+            return (type as? KtTypeWithNullability)?.nullability == KtTypeNullability.NON_NULLABLE
+        }
+
+        if (psiElement is KtDotQualifiedExpression) {
```

Comment:
Can we do these with:
```kt
when (psiElement) {
  is KtDotQualifiedExpression -> ...
  is KtNameReferenceExpression -> ...
  else -> ...
}
```
?

Reply:
Or, perhaps intentional, since the 2nd case is not sure yet?

Reply:
Done. The 2nd branch will always be necessary since it can be (mistakenly) reported on some other kind of expression